datafusion/sqllogictest/test_files/cse.slt

```diff
@@ -171,3 +175,41 @@ logical_plan
 physical_plan
 01)ProjectionExec: expr=[a@0 = random() AND b@1 = 0 as c1, a@0 = random() AND b@1 = 1 as c2, a@0 = 2 + random() OR b@1 = 4 as c3, a@0 = 2 + random() OR b@1 = 5 as c4, CASE WHEN a@0 = 4 + random() THEN 0 ELSE 1 END as c5, CASE WHEN a@0 = 4 + random() THEN 0 ELSE 2 END as c6]
 02)--MemoryExec: partitions=1, partition_sizes=[0]
+
```

Comment:
Could we maybe add some negative tests if they aren't already handled

For example, I think these should not be CSE'd:

```sql
    (random() = 0 OR a = 1) AND a = 1
```

```sql
    (random() = 0 AND a = 1) OR a = 1
```

```sql
    CASE 
      WHEN a + 10 = 0 THEN 0 
      WHEN random() > 0.5 THEN a+10 
      ELSE 0
    END
```

```sql
    CASE 
      WHEN random() > 0.5 THEN 0
      WHEN a + 10 = 0 THEN 0 
      ELSE a + 10
    END
```

```sql
    CASE 
      WHEN a + 10 = 0 THEN 0 
      WHEN random() > 0.5 
      WHEN random() > 0.5 THEN a+10 
      ELSE 0 
END
```

Reply:
This is a good idea! I forgot about negative tests. But from the first and 3rd case when examples above we can extract `a + 10` as it appears in the first when, so it is surely executed, and it also appear in the conditional subtrees.
I've added new tests in https://github.com/apache/datafusion/pull/11357/commits/6e3300f4416f1dfb4beb6d6fffb10a631447a826.